e3nn/SO3.py

```diff
@@ -376,13 +406,13 @@ def tensor_productRs(Rs_1, Rs_2, get_l_output=selection_rule):
                 C = clebsch_gordan(l, l_1, l_2, cached=True) * (2 * l + 1) ** 0.5
                 I = torch.eye(mul_1 * mul_2).view(mul_1 * mul_2, mul_1, mul_2)
                 m = torch.einsum("wuv,kij->wkuivj", I, C).view(dim_out, dim_1, dim_2)
-                mixing_matrix[index_out:index_out+dim_out, index_1:index_1+dim_1, index_2:index_2+dim_2] = m
+                Q[index_out:index_out+dim_out, index_1:index_1+dim_1, index_2:index_2+dim_2] = m
                 index_out += dim_out
 
             index_2 += dim_2
         index_1 += dim_1
 
-    return Rs_out, mixing_matrix
+    return Rs_out, Q
 
 
 def elementwise_tensor_productRs(Rs_1, Rs_2, get_l_output=selection_rule):
```

Comment:
This function needs clarifying. Is it also computing an orthonormal change of basis? Could we also get a usage case?

Reply:
Yes it is an orthogonal change of basis!

`tensor_product = einsum('ijk,j,k->i', tensor_productRs(Rs1, Rs2)[1], input1, input2)`

Reply:
I disagree with calling it a change of basis only because typically that is interpreted as a rotation `einsum('ij,j->i', ...)`. Rather, it is a contraction of two indices into one index. You cannot turn `ki,i->k` independently -- you must contract over both `i` and `j` simultaneously. Perhaps we can just call it `clebsch_gordan_tensor`?

Reply:
Further clarification -- the `ijk` tensor in `ijk,j,k->i` is not a change of basis, but if you go from `ijk,j->ik` and then `ik,k->i` -- this intermediate tensor `ik` is a proper change of basis tensor.

Reply:
Gah! Sorry @bkmi -- all my comments are specifically for the tensor product and not the other mixing matrices which are indeed permutation matrices. My apologies!

Reply:
Regarding specifically about the elementwise tensor product Rs function, I find the usage case very helpful in understanding its purpose.

And it is a change of basis in the representation right? i.e. from a representation of multiple spherical harmonics of order Rs_1 and Rs_2 to Rs_total (irreducible)

Reply:
It's a tensor product of two (independent) representation spaces into one. I think basis typically means you are staying "within the same space".

Reply:
I'm pretty sure it is a change of basis. `R1 \otimes R2` is a **reducible** representation and with this particular change of basis you make it block diagonal.

Reply:
I definitely agree that it is the conversion of a [reducible representation to a irreducible representation](https://en.wikipedia.org/wiki/Clebsch%E2%80%93Gordan_coefficients#Tensor_product_space). I also understand why one could call it a change of basis -- we are just projecting two "kets" and onto one "bra". But, I feel that it might be confusing for other folks as to what it means to do the tensor product.

Reply:
I think I would feel best if the matrix was explicitly called `clebsch_gordan_tensor` or something along those lines. That feels most specific and google-able for the general user.

Reply:
I renamed it `clebsch_gordan_tensor`